cdc/scheduler/schedule_dispatcher.go

```diff
@@ -0,0 +1,626 @@
+	checkpointTs model.Ts,
+) *BaseScheduleDispatcher {
+	// logger is just the global logger with the `changefeed-id` field attached.
+	logger := log.L().With(zap.String("changefeed-id", changeFeedID))
+
+	return &BaseScheduleDispatcher{
+		tables:               util.NewTableSet(),
+		captureStatus:        map[model.CaptureID]*captureStatus{},
+		moveTableManager:     newMoveTableManager(),
+		balancer:             newTableNumberRebalancer(logger),
+		changeFeedID:         changeFeedID,
+		logger:               logger,
+		communicator:         communicator,
+		checkpointTs:         checkpointTs,
+		lastTickCaptureCount: captureCountUninitialized,
+	}
+}
+
+type captureStatus struct {
+	// SyncStatus indicates what we know about the capture's internal state.
+	// We need to know this before we can make decision whether to
+	// dispatch a table.
+	SyncStatus captureSyncStatus
+
+	// Watermark fields
+	CheckpointTs model.Ts
+	ResolvedTs   model.Ts
+	// We can add more fields here in the future, such as: RedoLogTs.
+}
+
+type captureSyncStatus int32
+
+const (
+	// captureUninitialized indicates that we have not sent an `Announce` to the capture yet,
+	// nor has it sent us a `Sync` message.
+	captureUninitialized = captureSyncStatus(iota) + 1
+	// captureSyncSent indicates that we have sent a `Sync` message to the capture, but have had
+	// no response yet.
+	captureSyncSent
+	// captureSyncFinished indicates that the capture has been fully initialized and is ready to
+	// accept `DispatchTable` messages.
+	captureSyncFinished
+)
+
+// Tick implements the interface ScheduleDispatcher.
+func (s *BaseScheduleDispatcher) Tick(
+	ctx context.Context,
+	checkpointTs model.Ts,
+	// currentTables are tables that SHOULD be running given the current checkpoint-ts.
+	// It is maintained by the caller of this function.
+	currentTables []model.TableID,
+	captures map[model.CaptureID]*model.CaptureInfo,
+) (newCheckpointTs, resolvedTs model.Ts, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Update the internal capture list with information from the Owner
+	// (from Etcd in the current implementation).
+	s.captures = captures
+
+	// Checks for checkpoint regression as a safety measure.
+	if s.checkpointTs > checkpointTs {
+		s.logger.Panic("checkpointTs regressed",
+			zap.Uint64("old", s.checkpointTs),
+			zap.Uint64("new", checkpointTs))
+	}
+	// Updates the internally maintained last checkpoint-ts.
+	s.checkpointTs = checkpointTs
+
+	// Makes sure that captures have all been synchronized before proceeding.
+	done, err := s.syncCaptures(ctx)
+	if err != nil {
+		return CheckpointCannotProceed, CheckpointCannotProceed, errors.Trace(err)
+	}
+	if !done {
+		// Returns early if not all captures have synced their states with us.
+		// We need to know all captures' status in order to proceed.
+		// This is crucial for ensuring that no table is double-scheduled.
+		return CheckpointCannotProceed, CheckpointCannotProceed, nil
```

Comment:
do we have to add a log here to indicate that not all captured synced yet?

Reply:
There can be too many log messages, especially during a delta scan. There can easily be several messages per second per changefeed.